src/Moq/MatcherFactory.cs

```diff
@@ -67,20 +68,20 @@ public static IMatcher CreateMatcher(Expression argument, ParameterInfo paramete
 						throw new NotSupportedException(Resources.RefExpressionMustBeConstantValue);
 					}
 
-					return new RefMatcher(constant.Value);
+					return new Pair<IMatcher, Expression>(new RefMatcher(constant.Value), constant);
 				}
 			}
 			else if (parameter.IsDefined(typeof(ParamArrayAttribute), true) && (argument.NodeType == ExpressionType.NewArrayInit || !argument.Type.IsArray))
 			{
-				return new ParamArrayMatcher((NewArrayExpression)argument);
+				return new Pair<IMatcher, Expression>(new ParamArrayMatcher((NewArrayExpression)argument), argument);
```

Comment:
Using `argument` as the associated expression here isn't actually good enough, since `ParamArrayMatcher` will re-invoke `MatcherFactory` for each of its sub-parameter expressions. A new `NewArrayExpression` should be constructed from that.